### PR TITLE
Reworked `space_ops.py`

### PR DESCRIPTION
**Pull request by icedcoffeeee**
### List of Changes
- use scipy for rotations
- rewrite the functions to use the rotation matrix from scipy
- fix some typings
- rewrite z_to_vector to use cross products instead manually calculating using angles
- removed unnecessary functions (project_along_vector) and if statements
- rewrite line_intersection to use cleaner implementation
- added some more documentation
## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

### CODE REVIEW

File: manim/utils/space_ops.py [hunk 1]
```diff
@@ -582,19 +527,22 @@ def line_intersection(line1: Sequence[float], line2: Sequence[float]) -> np.ndar
...
-    div = det(x_diff, y_diff)
-    if div == 0:
-        raise ValueError("Lines do not intersect")
-    d = (det(*line1), det(*line2))
-    x = det(d, x_diff) / div
-    y = det(d, y_diff) / div
-    return np.array([x, y, 0])
+    if any(np.array([line1, line2])[:, :, 2].reshape(-1)):
+        # checks for z coordinates != 0
+        raise ValueError("Intersection for coords in the xy-plane.")
```
Comment (Kiran-Raj-Dev) [inline manim/utils/space_ops.py:532]:
```suggestion
        raise ValueError("Coords must be in the x-y plane.")
```
I think this sounds better

**Comment by Darylgolden** [comment]
I think you can just rewrite the line3d test not to have intersecting lines.

**Comment by icedcoffeeee** [comment]
The function `.perpendicular_to` creates the intersection though, so testing it will involve some intersecting lines.

**Comment by Darylgolden** [comment]
You can use normal lines for that.

**Comment by Darylgolden** [comment]
Could you add tests for the functions you've changed?

**Comment by icedcoffeeee** [comment]
> Could you add tests for the functions

Yep, sure.

Regarding the quaternions, on more digging, it might just be better if we fully use matrices for the opengl camera rotation.

**Comment by Darylgolden** [comment]
> > Could you add tests for the functions
> 
> Yep, sure.
> 
> Regarding the quaternions, on more digging, it might just be better if we fully use matrices for the opengl camera rotation. 

Maybe leave that for a different PR? There's already quite a lot here to review.

**Comment by icedcoffeeee** [comment]
> leave that for a different PR

Sure, i'll just revert the updates on those functions as well as the new dependency

File: tests/test_space_ops.py [hunk 2]
```diff
@@ -2,14 +2,87 @@
 
 import numpy as np
 
-from manim import cartesian_to_spherical, spherical_to_cartesian
+from manim.utils.space_ops import *
+from manim.utils.space_ops import shoelace, shoelace_direction
+
+
+def test_rotate_vector():
```
Comment (Darylgolden) [inline tests/test_space_ops.py:9]: Can you test the edge case where the vector is the zero vector?

File: tests/test_space_ops.py [hunk 3]
```diff
@@ -2,14 +2,87 @@
...
+            [
+                [0.96362, 0.0, 0.26726],
+                [-0.14825, 0.83205, 0.53452],
+                [-0.22237, -0.5547, 0.80178],
+            ]
+        )
+    )
+
+
+def test_angle_of_vector():
```
Comment (Darylgolden) [inline tests/test_space_ops.py:55]: Can you test the edge case where the vector is the zero vector?

File: tests/test_space_ops.py [hunk 4]
```diff
@@ -2,14 +2,87 @@
...
+        np.round(angle_between_vectors(np.array([1, 1, 1]), np.array([-1, 1, 1])), 5)
+        == 1.23096
+    )
+
+
+def test_center_of_mass():
+    assert np.all(center_of_mass([[0, 0, 0], [1, 2, 3]]) == np.array([0.5, 1.0, 1.5]))
+
+
+def test_line_intersection():
```
Comment (Darylgolden) [inline tests/test_space_ops.py:67]: Can you test that
- a ValueError is properly raised when the lines don't line in the xy plane;
- the case where the lines are parallel
- the case where the lines aren't parallel, but the intersection occurs outside the specified line segments

**Review by Darylgolden — CHANGES_REQUESTED** [review]
LGTM in general, just a few comments about testing edge case.

File: tests/test_space_ops.py (hunk 4 above)
Comment (icedcoffeeee) [reply tests/test_space_ops.py:67]: > the case where the lines aren't parallel, but the intersection occurs outside the specified line segments

This is not implemented, and neither was the previous implementation. Tbh, idt it should be.

Comment (Darylgolden) [reply tests/test_space_ops.py:70]: Hmm, what if the lines are equal too?

Comment (icedcoffeeee) [reply tests/test_space_ops.py:70]: Its pretty much the same as parallel lines, do you think we need to test for it too?

Comment (Darylgolden) [reply tests/test_space_ops.py:70]: I think more tests are (almost) never a bad thing

File: manim/utils/space_ops.py [hunk 5]
```diff
@@ -561,14 +507,17 @@ def midpoint(
     return center_of_mass([point1, point2])
 
 
-def line_intersection(line1: Sequence[float], line2: Sequence[float]) -> np.ndarray:
+def line_intersection(
+    line1: Sequence[np.ndarray], line2: Sequence[np.ndarray]
+) -> np.ndarray:
     """Returns intersection point of two lines, each defined with
```
Comment (Darylgolden) [inline manim/utils/space_ops.py:513]:
```suggestion
    """Returns intersection point of two line segments, each defined with
```

File: manim/utils/space_ops.py [hunk 6]
```diff
@@ -561,14 +507,17 @@ def midpoint(
...
     a pair of vectors determining the end points.
 
     Parameters
     ----------
     line1
-        The first line.
+        The first line, a list of two points representing the start
```
Comment (Darylgolden) [inline manim/utils/space_ops.py:519]:
```suggestion
        The first line segment, a list of two points representing the start
```

File: manim/utils/space_ops.py [hunk 7]
```diff
@@ -561,14 +507,17 @@ def midpoint(
...
+        and end.
     line2
         The second line.
```
Comment (Darylgolden) [inline manim/utils/space_ops.py:522]:
```suggestion
        The second line segment.
```

File: manim/utils/space_ops.py [hunk 8]
```diff
@@ -581,49 +530,47 @@ def line_intersection(line1: Sequence[float], line2: Sequence[float]) -> np.ndar
...
+        np.pad(np.array(i)[:, :2], ((0, 0), (0, 1)), constant_values=1)
+        for i in (line1, line2)
+    )
+    line1, line2 = (np.cross(*i) for i in padded)
+    x, y, z = np.cross(line1, line2)
 
-    def det(a, b):
-        return a[0] * b[1] - a[1] * b[0]
+    if z == 0:
+        raise ValueError("Lines do not intersect.")
```
Comment (Darylgolden) [inline manim/utils/space_ops.py:548]:
```suggestion
        raise ValueError("Line segments do not intersect.")
```

**Review by Darylgolden — CHANGES_REQUESTED** [review]
Line segment is the more technically correct term.

File: manim/utils/space_ops.py (hunk 8 above)
Comment (icedcoffeeee) [reply manim/utils/space_ops.py:548]: The rest seem appropriate, but this one implies that the segments themselves not intersecting would raise this error which isn't true. How about "The line segments are parallel." since it's the only time the error appears?

Comment (Darylgolden) [reply manim/utils/space_ops.py:548]: I thought from your previous comment, the function does not find a result if the intersection point is not within the line segments?

Comment (Kiran-Raj-Dev) [reply manim/utils/space_ops.py:548]: > The rest seem appropriate, but this one implies that the segments themselves not intersecting would raise this error which isn't true. How about "The line segments are parallel." since it's the only time the error appears?

Even if they are coincident not just parallel

Comment (icedcoffeeee) [reply manim/utils/space_ops.py:548]: > I thought from your previous comment, the function does not find a result if the intersection point is not within the line segments

I must've said something wrong, my bad. Currently, the function acts as if the two points given for each 'line' are points along an infinite line. Therefore, only if the lines are parallel (or coincident), the error is raised.

The complete message could be: "No intersections found. The lines are parallel or coincident."

Comment (Darylgolden) [reply manim/utils/space_ops.py:548]: Sounds good, except mentioning coincidence is redundant since coincident lines are just a special case of parallel lines.

**Review by Darylgolden — APPROVED** [review]
Thank you, this LGTM! (I haven't looked at the change in the graphical unit test control data, so if someone wants to double check that, it would be great)

**Comment by Kiran-Raj-Dev** [comment]
I liked `Lines do not intersect` better. Simple and to the point.

**Comment by Darylgolden** [comment]
> I liked `Lines do not intersect` better. Simple and to the point.

That's not really accurate for coincident lines though. How about "Lines are parallel"?

**Comment by Kiran-Raj-Dev** [comment]
But coincident lines also raise an error.
How about just `No Intersections found` ?

**Review by behackl — APPROVED** [review]
I've reviewed this coarsely and like all of these changes a lot. This is a gigantic improvement to `space_ops`, thank you!

**Comment by behackl** [comment]
My 2 cents regarding the parallel lines: I'd put "Lines are parallel, they do not have a unique intersection point."

File: manim/utils/space_ops.py [hunk 9]
```diff
@@ -561,16 +507,19 @@ def midpoint(
     return center_of_mass([point1, point2])
 
 
-def line_intersection(line1: Sequence[float], line2: Sequence[float]) -> np.ndarray:
-    """Returns intersection point of two lines, each defined with
+def line_intersection(
+    line1: Sequence[np.ndarray], line2: Sequence[np.ndarray]
+) -> np.ndarray:
+    """Returns intersection point of two line segments, each defined with
```
Comment (Kiran-Raj-Dev) [inline manim/utils/space_ops.py:513]:
```suggestion
    """Returns intersection point of two lines, each defined with
```
The formulas actually take into account the line defined by the vectors, and not the lne segment.
For Example:
Test_ManimCE_v0 14 0
The above line segments do not intersect but the lines do:
```py
>>> print(line_intersection([[-1,0,0], [0,3,0]],[[1,2,0], [-1,3,0]]))
[-0.14285714  2.57142857  0.        ] 
```

File: manim/utils/space_ops.py [hunk 10]
```diff
@@ -561,16 +507,19 @@ def midpoint(
...
     a pair of vectors determining the end points.
```
Comment (Kiran-Raj-Dev) [inline manim/utils/space_ops.py:514]:
```suggestion
    a pair of vectors determined by the lines passed as arguments
```
I think this better conveys the idea.

Comment (behackl) [inline manim/utils/space_ops.py:514]: An improved version of @Kiran-Raj-Dev's suggestions.

```suggestion
    """Returns the intersection point of two lines, each defined by
    a pair of distinct points lying on the line.
```

File: manim/utils/space_ops.py [hunk 11]
```diff
@@ -561,16 +507,19 @@ def midpoint(
...
 
     Parameters
     ----------
     line1
-        The first line.
+        The first line segment, a list of two points representing the start
+        and end.
```
Comment (behackl) [inline manim/utils/space_ops.py:520]:
```suggestion
        A list of two points that determine the first line.
```

File: manim/utils/space_ops.py [hunk 12]
```diff
@@ -561,16 +507,19 @@ def midpoint(
...
     line2
-        The second line.
+        The second line segment.
```
Comment (behackl) [inline manim/utils/space_ops.py:522]:
```suggestion
        A list of two points that determine the second line.
```

File: manim/utils/space_ops.py [hunk 13]
```diff
@@ -581,49 +530,47 @@ def line_intersection(line1: Sequence[float], line2: Sequence[float]) -> np.ndar
...
+        np.pad(np.array(i)[:, :2], ((0, 0), (0, 1)), constant_values=1)
+        for i in (line1, line2)
+    )
+    line1, line2 = (np.cross(*i) for i in padded)
+    x, y, z = np.cross(line1, line2)
 
-    def det(a, b):
-        return a[0] * b[1] - a[1] * b[0]
+    if z == 0:
+        raise ValueError("No intersections found. The lines are parallel.")
```
Comment (behackl) [inline manim/utils/space_ops.py:548]:
```suggestion
        raise ValueError("The lines are parallel, there is no unique intersection point.")
```

**Review by behackl — COMMENTED** [review]
Just some minor wording changes, will apply and then merge if nobody is opposed.

File: manim/utils/space_ops.py (hunk 10 above)
Comment (behackl) [reply manim/utils/space_ops.py]: Resolved by the rewording I just committed

File: manim/utils/space_ops.py (hunk 9 above)
Comment (behackl) [reply manim/utils/space_ops.py]: resolved by latest commit
